### PR TITLE
fix: [DEL-4402]: Making delegate xmx configurable based on POD's memory

### PR DESCRIPTION
**Pull request by gaurav-jain1**
## Describe your changes
Changes to make delegate process memory configurable based on container's RAM, this change should be applicable for image based delegates only.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/38127)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/38127)

**Review by AnupamIO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38127)* status: 3 file(s) to review. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 1329 at r1](https://reviewable.io/reviews/harness/harness-core/38127#-ND9LNYn8Co-ZEWwQlQi:-ND9LNYn8Co-ZEWwQlQj:b3kpyhi) ([raw file](https://github.com/harness/harness-core/blob/45a4a01e89b78cf916103568a276368bca623887/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L1329)):*
> ```Java
>             .logStreamingServiceBaseUrl(mainConfiguration.getLogStreamingServiceConfig().getBaseUrl())
>             .delegateTokenName(delegateTokenName.orElse(null))
>             .watcher(true)
> ```

Could you check if we really need to set this watcher true. Previously it was needed to get JreConfig but it was removed in Recent PR by Arpit: https://github.com/harness/harness-core/pull/37960

**Review by AnupamIO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38127)* status: 3 file(s) to review. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 4370 at r1](https://reviewable.io/reviews/harness/harness-core/38127#-ND9No1uDuB9E-3U4JIh:-ND9No1vAyBFtaY08tVb:b-ktb3mb) ([raw file](https://github.com/harness/harness-core/blob/45a4a01e89b78cf916103568a276368bca623887/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L4370)):*
> ```Java
>       return "-Xmx1536m";
>     }
>     return "-XX:MaxRAMPercentage=60.0 -XX:MinRAMPercentage=30.0";
> ```

asking to understand: Why are we keeping it between 30-60% ?

**Review by AnupamIO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38127)* status: 3 file(s) to review. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 1567 at r1](https://reviewable.io/reviews/harness/harness-core/38127#-ND9OfdO-xWRzlzExjVj:-ND9OfdO-xWRzlzExjVk:bs4co7) ([raw file](https://github.com/harness/harness-core/blob/44f8a49918798098e8ce43e62717f296dd4e2c69/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L1567)):*
> ```Java
>       params.put("delegateXmx", templateParameters.getDelegateXmx());
>     } else {
>       if (featureFlagService.isEnabled(REDUCE_DELEGATE_MEMORY_SIZE, templateParameters.getAccountId())) {
> ```

REDUCE_DELEGATE_MEMORY_SIZE do we actually use this FF ?

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38127)* status: 3 file(s) to review. 4. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 1329 at r1](https://reviewable.io/reviews/harness/harness-core/38127#-ND9LNYn8Co-ZEWwQlQi:-NDC4aYFANTVt7G7_Jdy:b-jufmkm) ([raw file](https://github.com/harness/harness-core/blob/45a4a01e89b78cf916103568a276368bca623887/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L1329)):*
> eck if we really need to set this watcher true. Previously it was needed to get JreConfig but it 
Yeah, we can take that clean up separately. created https://harness.atlassian.net/browse/DEL-5011 to track
___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 1567 at r1](https://reviewable.io/reviews/harness/harness-core/38127#-ND9OfdO-xWRzlzExjVj:-NDC5JFe9t-MqSxyKCw7:bii2yez) ([raw file](https://github.com/harness/harness-core/blob/44f8a49918798098e8ce43e62717f296dd4e2c69/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L1567)):*
<details><summary><i>Previously, AnupamIO (Anupam Jaiswal) wrote…</i></summary><blockquote>

REDUCE_DELEGATE_MEMORY_SIZE do we actually use this FF ?
</blockquote></details>

Not sure if we are using it now @raghuatwings or @bharatgoelharness  can confirm
___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 4370 at r1](https://reviewable.io/reviews/harness/harness-core/38127#-ND9No1uDuB9E-3U4JIh:-NDC5VOgDc4ybYVLQ_LH:b6ql41r) ([raw file](https://github.com/harness/harness-core/blob/45a4a01e89b78cf916103568a276368bca623887/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L4370)):*
<details><summary><i>Previously, AnupamIO (Anupam Jaiswal) wrote…</i></summary><blockquote>

asking to understand: Why are we keeping it between 30-60% ?
</blockquote></details>

this calculated to current xmx value, i.e. 1500MB xmx with 2G PODs memory.

File: 400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java [hunk 1]
```diff
@@ -4370,10 +4361,13 @@ private boolean checkForDelegateHavingAllTags(Delegate delegate, DelegateTags ta
...
-    return (DeployVariant.isCommunity(deployVersion) || (delegateType != null && (delegateType.equals(DOCKER))))
-        ? "-Xmx512m"
-        : "-Xmx1536m";
+    if (DeployVariant.isCommunity(deployVersion)) {
+      return "-Xmx512m";
+    }
+    if ((delegateType != null && (delegateType.equals(SHELL_SCRIPT)))) {
+      return "-Xmx1536m";
+    }
+    return "-XX:MaxRAMPercentage=60.0 -XX:MinRAMPercentage=30.0";
```
Comment (xingchi-jin) [inline 400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java:4370]: where does the Max and Min ram come from ?

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38127)* status: Some of the checks are too old. 3 file(s) to review. 5. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 4370 at r1](https://reviewable.io/reviews/harness/harness-core/38127#-ND7cvJGYe6y8sJv0bwl-r1-4370:-NEisMAx26b3E9HkIeSp:b-mpffnj) ([raw file](https://github.com/harness/harness-core/blob/45a4a01e89b78cf916103568a276368bca623887/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L4370)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

where does the Max and Min ram come from ? 
</blockquote></details>

this came from historic data, current pod's size.

**Review by mdmilic — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38127)* status: Some of the checks are too old. 1 change requests. 5. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 4365 at r1](https://reviewable.io/reviews/harness/harness-core/38127#-NFNykfm0mUcMG3h2BAK:-NFNykfm0mUcMG3h2BAL:b-ol6xjb) ([raw file](https://github.com/harness/harness-core/blob/45a4a01e89b78cf916103568a276368bca623887/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L4365)):*
> ```Java
>   private String getDelegateXmx(String delegateType) {
>     if (DeployVariant.isCommunity(deployVersion)) {
>       return "-Xmx512m";
> ```

Why do we need to keep XMX for community users? Also, Arpit's comment was suggesting that it should be removed. 
___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 4367 at r1](https://reviewable.io/reviews/harness/harness-core/38127#-NFNyICIDlmBGZYH5GZA:-NFNyICIDlmBGZYH5GZB:b-kh6608) ([raw file](https://github.com/harness/harness-core/blob/45a4a01e89b78cf916103568a276368bca623887/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L4367)):*
> ```Java
>       return "-Xmx512m";
>     }
>     if ((delegateType != null && (delegateType.equals(SHELL_SCRIPT)))) {
> ```

You can simplify this if statement by inverting the equals



_Suggestion:_
```Java
if (SHELL_SCRIPT.equals(delegateType)) {
```

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38127)* status: Some of the checks are too old. 1 change requests. 6. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java` line 4365 at r1](https://reviewable.io/reviews/harness/harness-core/38127#-NFNykfm0mUcMG3h2BAK:-NFRxTVcDYcnYRIFWQv1:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/45a4a01e89b78cf916103568a276368bca623887/400-rest/src/main/java/software/wings/service/impl/DelegateServiceImpl.java#L4365)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Why do we need to keep XMX for community users? Also, Arpit's comment was suggesting that it should be removed. 
</blockquote></details>

Done.

**Review by mdmilic — APPROVED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38127)* status: Some of the checks are too old. 5. unresolved discussions.

**Pull request merged**